demos/http/http_demo_s3_download_async/http_demo_s3_download_async.c

```diff
@@ -0,0 +1,751 @@
+                /* Log error to indicate connection failure after all
+                 * reconnect attempts are over. */
+                LogError( ( "Failed to connect to HTTP server %s.",
+                            serverHost ) );
+            }
+        }
+
+        /* Define the transport interface. */
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            transportInterface.recv = Openssl_Recv;
+            transportInterface.send = Openssl_Send;
+            transportInterface.pNetworkContext = &networkContext;
+        }
+
+        /******************** Start queues and HTTP task. *******************/
+
+        /* Start request and response queues */
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            queueSettings.mq_maxmsg = 10;
+            queueSettings.mq_msgsize = sizeof( RequestItem_t );
+
+            requestQueue = mq_open( REQUEST_QUEUE,
+                                    O_CREAT | O_WRONLY,
+                                    0700,
+                                    &queueSettings );
+
+            queueSettings.mq_msgsize = sizeof( ResponseItem_t );
+
+            responseQueue = mq_open( RESPONSE_QUEUE,
+                                     O_CREAT | O_RDONLY,
+                                     0700,
+                                     &queueSettings );
+
+            if( requestQueue == -1 )
+            {
+                LogError( ( "Failed to open request queue with error %s.",
+                            strerror( errno ) ) );
+                returnStatus = EXIT_FAILURE;
+            }
+
+            if( responseQueue == -1 )
+            {
+                LogError( ( "Failed to open response queue with error %s.",
+                            strerror( errno ) ) );
+                returnStatus = EXIT_FAILURE;
+            }
+        }
+
+        /* Start HTTP task */
+
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            httpThread = startHTTPThread( &transportInterface );
+
+            if( httpThread == -1 )
+            {
+                returnStatus = EXIT_SUCCESS;
+            }
+        }
+
+        /******************** Download S3 Object File. **********************/
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            returnStatus = requestS3ObjectFile( serverHost,
+                                                HTTP_METHOD_GET,
+                                                pPath,
+                                                requestQueue,
+                                                &requestCount );
+        }
+
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            retrieveS3ObjectFile( responseQueue, &requestCount );
+        }
+
+        /************************** Disconnect. *****************************/
+
+        /* End TLS session, then close TCP connection. */
+        ( void ) Openssl_Disconnect( &networkContext );
+
+        /******************** Clean up queues and HTTP task. ****************/
```

Comment:
Having a tearDown() function for this would make for more readability.

Reply:
Yea I also think it would be good to put this queue closing stuff in a separate function